clang/lib/CIR/CodeGen/TargetInfo.cpp

```diff
@@ -159,6 +218,66 @@ class X86_64TargetCIRGenInfo : public TargetCIRGenInfo {
 };
 } // namespace
 
+//===----------------------------------------------------------------------===//
+// Base ABI and target codegen info implementation common between SPIR and
+// SPIR-V.
+//===----------------------------------------------------------------------===//
+
+namespace {
+class CommonSPIRABIInfo : public DefaultABIInfo {
+public:
+  CommonSPIRABIInfo(CIRGenTypes &CGT) : DefaultABIInfo(CGT) {}
+};
+
+class SPIRVABIInfo : public CommonSPIRABIInfo {
+public:
+  SPIRVABIInfo(CIRGenTypes &CGT) : CommonSPIRABIInfo(CGT) {}
+  void computeInfo(CIRGenFunctionInfo &FI) const override {
+    // The logic is same as in DefaultABIInfo with an exception on the kernel
+    // arguments handling.
+    llvm::CallingConv::ID CC = FI.getCallingConvention();
+
+    bool cxxabiHit = getCXXABI().classifyReturnType(FI);
+    assert(!cxxabiHit && "C++ ABI not considered");
+
+    FI.getReturnInfo() = classifyReturnType(FI.getReturnType());
+
+    for (auto &I : FI.arguments()) {
+      if (CC == llvm::CallingConv::SPIR_KERNEL) {
+        I.info = classifyKernelArgumentType(I.type);
+      } else {
+        I.info = classifyArgumentType(I.type);
+      }
+    }
+  }
+
+private:
+  ABIArgInfo classifyKernelArgumentType(QualType Ty) const {
+    assert(!getContext().getLangOpts().CUDAIsDevice && "NYI");
+    return classifyArgumentType(Ty);
+  }
+};
+} // namespace
+namespace {
+
+class CommonSPIRTargetCIRGenInfo : public TargetCIRGenInfo {
+public:
+  CommonSPIRTargetCIRGenInfo(std::unique_ptr<ABIInfo> ABIInfo)
+      : TargetCIRGenInfo(std::move(ABIInfo)) {}
+
```

Comment:
Let's clarify the absence of `ASTAllocaAddressSpace` in this PR.

All allocas in OpenCL reside in the "private" address space, which is the default(`0`) and does not requires any casting. Clang maps both `LangAS::Default` and `LangAS::FirstTargetAddressSpace` to `0` as "target address space", which fills the `addrspace(●)` in LLVM IR.

https://github.com/llvm/clangir/blob/9720c614b93406c6a08a04887958c2e0e3ad1632/clang/lib/CodeGen/Targets/SPIR.cpp#L46-L49

In original CodeGen SPIR-V target, `getLangASFromTargetAS` is used for the reversed mapping. As a result, it returns `FirstTargetAddressSpace`, which is mistakenly treated as "non-default" alloca address space.

Therefore, actually SPIR-V target itself does not depend on "support non-default alloca address space". In other words, leaving the method not overrided and returning `LangAS::Default` as fallback should be already a correct implementation.

Moreover, since there is only one target address space for alloca, we could not construct any valid test cases for `ASTAllocaAddressSpace` in SPIR-V: different `LangAS`s are not observable in the final CIR or LLVM IR.

I suggest to improve those checks that wrongly treat `FirstTargetAddressSpace` as non-default (like, uses target address spaces for casting comparison). Then we can add the override back. So it's not included in this PR. I will tackle it soon.

@Naghasan Wdyt?

Reply:
See also #672.

Reply:
Thanks for detailed explanation.

> which is mistakenly treated as "non-default" alloca address space

Mistakenly how? I don't understand the technical reasons why it's hard to fix this right away.

> leaving the method not overrided and returning LangAS::Default as fallback should be already a correct implementation.

Being correct but deviating from OG codegen for no good reason doesn't sounds like a good rationale to me, it will just complicate things for folks touching this code in future. Perhaps you should consider breaking the pieces you need in a more fine grained way and adding them first - workaround like this tend to stick forever unless the right approach is done right away.

Btw, I didn't understand if you are planning to add `getASTAllocaAddressSpace` in the near future, please clarify.

Reply:
>> which is mistakenly treated as "non-default" alloca address space
> Mistakenly how? I don't understand the technical reasons why it's hard to fix this right away.

I believe @seven-mile refers to this checks:

---

**CIR**
https://github.com/llvm/clangir/blob/f07bad43ee1d82452b04a0fcdbd57e82ad8ef060/clang/lib/CIR/CodeGen/CIRGenExpr.cpp#L2946-L2947

**OG** https://github.com/llvm/clangir/blob/f07bad43ee1d82452b04a0fcdbd57e82ad8ef060/clang/lib/CodeGen/CGExpr.cpp#L96-L107

---

**CIR**
https://github.com/llvm/clangir/blob/f07bad43ee1d82452b04a0fcdbd57e82ad8ef060/clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp#L1002-L1005

**OG**
https://github.com/llvm/clangir/blob/f07bad43ee1d82452b04a0fcdbd57e82ad8ef060/clang/lib/CodeGen/CGBuiltin.cpp#L4103-L4109

---

The underlying problem (quirk?) is that the SPIR target returns a "target" `LangAS` instead of just `LangAS::Default`, hence the OG codegen attempts to create an addrspacecast. However, it turns out to be a nop because both are mapped to the same address space identifier `0` in LLVM-IR.

Thoughts:
- I think we should implement `getASTAllocaAddressSpace` to mimic the OG SPIR target right away.
- @seven-mile's proposal to refine the NYI-checks in CIR to look at the resulting target address space to determine if we actually would create a cast makes sense to me.
- On the other hand, we already have the CIR address space cast, so we could just use it.
- Testing remains an issue. @bcardosolopes would you be open to add a dummy target to exercise these code paths?

Reply:
> The underlying problem (quirk?) is that the SPIR target returns a "target" LangAS instead of just LangAS::Default, hence the OG codegen attempts to create an addrspacecast. However, it turns out to be a nop because both are mapped to the same address space identifier 0 in LLVM-IR.

I don't see how that's a problem, it's expected behavior that CIRGen might emit some dumb code that gets cleaned up later, e.g. `CastOp::fold`. Alternatively, if the cast operation builder is made a bit more smart, it could decide to return the source instead of emitting a new cast.

> I think we should implement getASTAllocaAddressSpace to mimic the OG SPIR target right away.

+1

> @seven-mile's proposal to refine the NYI-checks in CIR to look at the resulting target address space to determine if we actually would create a cast makes sense to me.

Incremental work is fine, the side effect here is that it might need incremental redesign of how we treat address space, since more constraints or behavior will show up as the incremental pieces are added.

It seems to me that we just hit one design point early in the game, which is the distinction between FirstTargetAddressSpace and default address space. This is a good indication this one needs to be solved right away.

> On the other hand, we already have the CIR address space cast, so we could just use it.

I'm fine with this as long as we keep refining the design.

> Testing remains an issue. @bcardosolopes would you be open to add a dummy target to exercise these code paths?

If SPIRV is what you are targetting as part of OpenCL work, I think we should just add the proper bits right away. But if by dummy you are talking abour other target, than we should add the minimal bits for that real target.

Reply:
Sorry for the late reply, I was off till now.

> It seems to me that we just hit one design point early in the game,

Seems that way, I think we need to make a call on which kind of address space we want in the dialect (LangAS, target address space or something in between).

>  which is the distinction between FirstTargetAddressSpace and default address space. This is a good indication this one needs to be solved right away.

Just restate the idea behind `LangAS`: it is a strongly type enum use to carry either a language address space (specific to a language and having a meaning for clang) or a target address space (a number only meaningful to the backend). In order to separate the two, target address spaces (simple numbers) are offset-ed by `FirstTargetAddressSpace`, So `FirstTargetAddressSpace` isn't mapped to anything, it is the `0` target address space in the `LangAS` encoding.

> In original CodeGen SPIR-V target, getLangASFromTargetAS is used for the reversed mapping. As a result, it returns FirstTargetAddressSpace, which is mistakenly treated as "non-default" alloca address space.

It is actually the intent, `Default` is either `0` or `4`  depending on the language / mode (OpenCL and SYCL). Treating it this way ensure you pick up the right address space for alloca in a SPIR target.

Reply:
> Seems that way, I think we need to make a call on which kind of address space we want in the dialect (LangAS, target address space or something in between).

I think I confused myself with earlier discussion with Julian